File: go/security/ble_privacy_fixture_test.go
// SPDX-License-Identifier: MIT

package security

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// Cross-language BLE tracking-protection fixture verifier. Reproduces the
// rotating Service UUID and the IRK-based Resolvable Private Address (RPA)
// against the canonical vectors in fixtures/bleprivacy/vectors.json and asserts
// they match the C# reference (and every other SDK) byte-for-byte.
//
// Any drift between this Go implementation and BlePrivacy.cs shows up here as a
// UUID string or hex mismatch.

type blePrivacyVectors struct {
	RotationSeconds int    `json:"rotation_seconds"`
	RotationKey     string `json:"rotation_key"`
	IRK             string `json:"irk"`
	WrongIRK        string `json:"wrong_irk"`
	UUIDVectors     []struct {
		Window int64  `json:"window"`
		UUID   string `json:"uuid"`
	} `json:"uuid_vectors"`
	RPAVectors []struct {
		Window int64  `json:"window"`
		RPA    string `json:"rpa"`
	} `json:"rpa_vectors"`
}

func loadBlePrivacyVectors(t *testing.T) blePrivacyVectors {
	t.Helper()
	root := repoRoot(t)
	path := filepath.Join(root, "fixtures", "bleprivacy", "vectors.json")

	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}
	var v blePrivacyVectors
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("unmarshal vectors: %v", err)
	}
	return v
}

func TestBlePrivacyFixture_ServiceUUID(t *testing.T) {
	v := loadBlePrivacyVectors(t)
	rotationKey := mustHex(t, v.RotationKey)

	if len(v.UUIDVectors) == 0 {
		t.Fatal("no uuid_vectors in fixture")
	}
	for _, uv := range v.UUIDVectors {
		got := ServiceUUID(rotationKey, uv.Window)
		if got != uv.UUID {
			t.Errorf("ServiceUUID(window=%d) mismatch:\n  expected: %s\n  actual:   %s",
				uv.Window, uv.UUID, got)
		}
	}
}

func TestBlePrivacyFixture_ResolvableAddress(t *testing.T) {
	v := loadBlePrivacyVectors(t)
	irk := mustHex(t, v.IRK)
	wrongIRK := mustHex(t, v.WrongIRK)

	if len(v.RPAVectors) == 0 {
		t.Fatal("no rpa_vectors in fixture")
	}
	for _, rv := range v.RPAVectors {
		rpa, err := ResolvableAddress(irk, rv.Window)
		if err != nil {
			t.Fatalf("ResolvableAddress(window=%d): %v", rv.Window, err)
		}
		if got := hex.EncodeToString(rpa); got != rv.RPA {
			t.Errorf("ResolvableAddress(window=%d) mismatch:\n  expected: %s\n  actual:   %s",
				rv.Window, rv.RPA, got)
		}

		// The generating IRK resolves its own RPA; a different IRK must not.
		if !ResolveAddress(irk, rpa) {
			t.Errorf("ResolveAddress(irk, rpa) = false for window=%d, want true", rv.Window)
		}
		if ResolveAddress(wrongIRK, rpa) {
			t.Errorf("ResolveAddress(wrongIRK, rpa) = true for window=%d, want false", rv.Window)
		}
	}
}

func TestBlePrivacyFixture_RotationSecondsAndWindowFor(t *testing.T) {
	v := loadBlePrivacyVectors(t)

	if RotationSeconds != v.RotationSeconds {
		t.Errorf("RotationSeconds = %d, want %d", RotationSeconds, v.RotationSeconds)
	}
	if w := WindowFor(899); w != 0 {
		t.Errorf("WindowFor(899) = %d, want 0", w)
	}
	if w := WindowFor(900); w != 1 {
		t.Errorf("WindowFor(900) = %d, want 1", w)
	}
}

func TestBlePrivacyFixture_ShortIRKRejected(t *testing.T) {
	shortIRK := make([]byte, 15)

	if _, err := ResolvableAddress(shortIRK, 0); err == nil {
		t.Error("ResolvableAddress(15-byte IRK) = nil error, want error")
	}
	if ResolveAddress(shortIRK, make([]byte, 6)) {
		t.Error("ResolveAddress(15-byte IRK, ...) = true, want false")
	}
}

File: kotlin/src/test/kotlin/aethernet/security/BlePrivacyFixtureTest.kt
// SPDX-License-Identifier: MIT
package aethernet.security

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * Cross-language BLE-privacy fixture verifier.
 *
 * Verifies that the Kotlin implementation reproduces fixtures/bleprivacy/vectors.json
 * byte-for-byte — the same rotating Service UUID and IRK-based Resolvable Private
 * Address (RPA) outputs that the C# reference and every AetherNet language SDK must
 * produce. Any drift surfaces here as a UUID / hex mismatch.
 */
class BlePrivacyFixtureTest {

    // ─── Fixture parity gate ────────────────────────────────────────────────

    @Test
    fun `bleprivacy uuid vectors match byte-for-byte`() {
        val fx = loadFixture()
        assertTrue(fx.uuidVectors.isNotEmpty(), "no uuid_vectors in fixture")
        for (v in fx.uuidVectors) {
            assertEquals(
                v.text,
                BlePrivacy.serviceUuid(fx.rotationKey, v.window),
                "serviceUuid mismatch for window=${v.window}",
            )
        }
    }

    @Test
    fun `bleprivacy rpa vectors resolve and reject byte-for-byte`() {
        val fx = loadFixture()
        assertTrue(fx.rpaVectors.isNotEmpty(), "no rpa_vectors in fixture")
        for (v in fx.rpaVectors) {
            val rpa = BlePrivacy.resolvableAddress(fx.irk, v.window)
            assertEquals(v.text, hex(rpa), "resolvableAddress mismatch for window=${v.window}")

            assertTrue(
                BlePrivacy.resolveAddress(fx.irk, rpa),
                "correct IRK must resolve its own RPA (window=${v.window})",
            )
            assertFalse(
                BlePrivacy.resolveAddress(fx.wrongIrk, rpa),
                "wrong IRK must not resolve the RPA (window=${v.window})",
            )
        }
    }

    @Test
    fun `bleprivacy rotation seconds matches fixture`() {
        val fx = loadFixture()
        assertEquals(fx.rotationSeconds, BlePrivacy.ROTATION_SECONDS)
    }

    // ─── windowFor boundary + reject paths ──────────────────────────────────

    @Test
    fun `windowFor floors to the rotation boundary`() {
        assertEquals(0L, BlePrivacy.windowFor(899))
        assertEquals(1L, BlePrivacy.windowFor(900))
    }

    @Test
    fun `resolvableAddress rejects a 15-byte IRK`() {
        assertThrows<IllegalArgumentException> {
            BlePrivacy.resolvableAddress(ByteArray(15), 0L)
        }
    }

    // ─── Fixture model + loader (mirrors Bip39FixtureTest / SignalFixtureTest) ─

    private data class Vector(val window: Long, val text: String)

    private data class Fixture(
        val rotationSeconds: Int,
        val rotationKey: ByteArray,
        val irk: ByteArray,
        val wrongIrk: ByteArray,
        val uuidVectors: List<Vector>,
        val rpaVectors: List<Vector>,
    )

    private fun loadFixture(): Fixture {
        val json = File(repoRoot(), "fixtures/bleprivacy/vectors.json").readText()
        return Fixture(
            rotationSeconds = scalarInt(json, "rotation_seconds"),
            rotationKey = unhex(scalarString(json, "rotation_key")),
            irk = unhex(scalarString(json, "irk")),
            wrongIrk = unhex(scalarString(json, "wrong_irk")),
            uuidVectors = parseVectors(json, "uuid_vectors", "uuid"),
            rpaVectors = parseVectors(json, "rpa_vectors", "rpa"),
        )
    }

    private fun repoRoot(): File {
        var dir: File? = File(".").canonicalFile
        repeat(10) {
            val candidate = File(dir, "AetherNetProtocol.slnx")
            if (candidate.exists()) return dir!!
            dir = dir?.parentFile ?: return@repeat
        }
        throw IllegalStateException("AetherNetProtocol.slnx not found from ${File(".").canonicalFile}")
    }

    private fun scalarString(json: String, name: String): String =
        Regex("\"$name\"\\s*:\\s*\"([^\"]*)\"").find(json)!!.groupValues[1]

    private fun scalarInt(json: String, name: String): Int =
        Regex("\"$name\"\\s*:\\s*(-?\\d+)").find(json)!!.groupValues[1].toInt()

    /**
     * Parses a named array of `{ "window": <int>, "<textKey>": "<string>" }`
     * objects. Same bracket-scanning approach as the sibling fixture tests, but
     * pulls the numeric `window` alongside the string field per object.
     */
    private fun parseVectors(json: String, arrayName: String, textKey: String): List<Vector> {
        val nameIdx = json.indexOf("\"$arrayName\"")
        require(nameIdx >= 0) { "No '$arrayName' array in vectors.json" }
        val openBracket = json.indexOf('[', nameIdx)
        var depth = 0
        var endIdx = -1
        for (i in openBracket until json.length) {
            when (json[i]) {
                '[' -> depth++
                ']' -> { depth--; if (depth == 0) { endIdx = i; break } }
            }
        }
        require(endIdx > 0) { "Unbalanced '$arrayName' array" }
        val arrayBody = json.substring(openBracket + 1, endIdx)
        return splitObjects(arrayBody).map { obj ->
            val window = Regex("\"window\"\\s*:\\s*(-?\\d+)").find(obj)!!.groupValues[1].toLong()
            val text = Regex("\"$textKey\"\\s*:\\s*\"([^\"]*)\"").find(obj)!!.groupValues[1]
            Vector(window, text)
        }
    }

    private fun splitObjects(text: String): List<String> {
        val result = mutableListOf<String>()
        var depth = 0
        var start = -1
        for (i in text.indices) {
            when (text[i]) {
                '{' -> { if (depth == 0) start = i; depth++ }
                '}' -> { depth--; if (depth == 0 && start >= 0) { result += text.substring(start, i + 1); start = -1 } }
            }
        }
        return result
    }

    private fun unhex(s: String): ByteArray {
        require(s.length % 2 == 0) { "hex string must have even length" }
        return ByteArray(s.length / 2) { i ->
            ((Character.digit(s[i * 2], 16) shl 4) + Character.digit(s[i * 2 + 1], 16)).toByte()
        }
    }

    private fun hex(b: ByteArray): String =
        b.joinToString("") { "%02x".format(it.toInt() and 0xFF) }
}
